cmd/logger.go

```diff
@@ -106,38 +95,79 @@ func (log *Logger) Printf(format string, args ...interface{}) {
+func initTrimStrings() []string {
+	var goPathList []string
+	trimStrings := []string{
+		filepath.Join(runtime.GOROOT(), "src") + string(filepath.Separator),
 	}
+	// Add all possible GOPATH paths into trimStrings
+	// Split GOPATH path depending on the type of OS
+	fmt.Println("runtime.GOOS:", runtime.GOOS)
+	if runtime.GOOS == "windows" {
+		goPathList = strings.Split(GOPATH, ";")
+	} else {
+		// All other types of OSs
+		goPathList = strings.Split(GOPATH, ":")
+	}
+	for _, goPathString := range goPathList {
+		trimStrings = append(trimStrings, filepath.Join(goPathString, "src")+string(filepath.Separator))
+	}
+	// Add "github.com/minio/minio" as the last. This is supposed to be appended
+	// after GOPATH and GOROOT to be able to trim "{GOROOT}/github.com/minio/minio"
+	// and "{GOPATH}/github.com/minio/minio" correctly
+	trimStrings = append(trimStrings, filepath.Join("github.com", "minio", "minio")+string(filepath.Separator))
 
-	// Adds a console logger.
-	logger.Hooks.Add(l)
-
-	return l
+	return trimStrings
 }
 
-func getSource() string {
-	var funcName string
-	pc, filename, lineNum, ok := runtime.Caller(2)
-	if ok {
-		filename = path.Base(filename)
-		funcName = strings.TrimPrefix(runtime.FuncForPC(pc).Name(), "github.com/minio/minio/cmd.")
-	} else {
-		filename = "<unknown>"
-		lineNum = 0
+func trimTrace(f string, trimStrings []string) string {
+	for _, trimString := range trimStrings {
+		f = strings.TrimPrefix(filepath.ToSlash(f), filepath.ToSlash(trimString))
 	}
+	return filepath.FromSlash(f)
+}
 
-	return fmt.Sprintf("[%s:%d:%s()]", filename, lineNum, funcName)
+// getTrace method - creates and returns stack trace
+func getTrace(traceLevel int) []string {
+	var trace []string
+	trimStrings := initTrimStrings()
+	pc, file, lineNumber, ok := runtime.Caller(traceLevel)
+
+	for ok {
+		// Clean up the common prefixes
+		file = trimTrace(file, trimStrings)
+		// Get the function name
+		_, funcName := filepath.Split(runtime.FuncForPC(pc).Name())
+		// Skip duplicate traces that start with file name, "<autogenerated>"
+		// and also skip traces with function name that starts with "runtime."
+		if !strings.HasPrefix(file, "<autogenerated>") &&
+			!strings.HasPrefix(funcName, "runtime.") {
+			// Form and append a line of stack trace into a
+			// collection, 'trace', to build full stack trace
+			trace = append(trace, fmt.Sprintf("%v:%v:%v()", file, lineNumber, funcName))
+		}
+		traceLevel++
+		// Read stack trace information from PC
+		pc, file, lineNumber, ok = runtime.Caller(traceLevel)
+	}
+	// Add a sequence number and formatting for each stack trace
+	// if json format is not selected
+	if !log.json {
+		// No formatting is required for the first entry
+		trace[0] = "1: " + trace[0]
+		for index, element := range trace[1:] {
+			// First element's index is actually index + 1
+			// since we took care of the first entry before
+			index++
+			trace[index] = fmt.Sprintf("%8v: %s", index+1, element)
```

Comment:
It would be clear to use `index + 1` and `index + 2`.  To avoid confusion here, it would be better to do `for` like
```go
for i, element := range trace {
    format := "%8v: %v"
    if i == 0 {
        format = "%v: %v"
    }

    trace[i] = fmt.Sprintf(format, i+1, element)
}
```

IMO `getTrace()` shouldn't do these formatting.

Reply:
I deliberately kept the handling of the first, `trace[0]`, entry outside the `for` loop to stop having a check (`if i == 0`) done for each and every trace entry unnecessarily. I'd like to  keep it that way.

I agree formatting does not belong in `getTrace()` function. So, I moved the formatting block into the `logIf()` function where we were already doing log formatting. This also helped getting rid of the extra check for `json` mode in the current code, since we already have a `json` mode check inside the `logIf()` function.
Done.